{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sys import path\n",
    "from os.path import join as joinpath\n",
    "from os.path import normpath\n",
    "path.append(normpath(\"../src\"))\n",
    "\n",
    "from protein import loadFasta\n",
    "from align import Align, Aligned\n",
    "from score import PSSM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# PSSM Profiles and alignment\n",
    "\n",
    "## Profiles\n",
    "\n",
    "A *profile* is a way to represent multiple aligned sequences and their similarities or patterns. \n",
    "\n",
    "## Definitions\n",
    "\n",
    "A protein **domain** is part of a protein that has its own structure and function, independent of the rest of the protein. Despite their use being similar across the proteins they can be found in, the same domains do not necessarily have the same amino acid sequence. However, they do have enough common amino acids to be identified and located by the means of *sequence alignment*. Our goal in this project is to locate a given domain within a given sequence, should it exist one or more times within that sequence. One such domain is the WW domain, known to be used by multiple species and sometimes included multiple times in a single protein. In this project we will use sequences from the WW domain that belong to human proteins to test our implementations : these have been saved to the file `to-be-aligned.fasta`. These domain sequences and many others can be found on the [SMART](http://smart.embl.de) database.\n",
    "\n",
    "\n",
    "**Multiple Sequence Alignment** (or MSA) allows us to align multiple sequences together, and therefore observe how well some amino acids are preserved at some positions. This is of great interest in the study of domains, since it reveals which positions are of greater importance when trying to align new sequences to known domains. There are of course other applications, as well as numerous methods that achieve MSA, however we won't go into them since we won't be implementing any of these in this project. Two online tools were used to align all the sequences from `to-be-aligned.fasta` together : [MUSCLE](http://www.ebi.ac.uk/Tools/msa/muscle/) and [CLUSTAL Omega](http://www.ebi.ac.uk/Tools/msa/clustalo/). The resulting sequences can be respectively found in files `msaresults-MUSCLE.fasta` and `msaresults-CLUSTALO.fasta`.\n",
    "\n",
    "## PSSM\n",
    "\n",
    "A **Position Specific Scoring Matrix** (or PSSM) is a kind of profile that uses a matrix, where each **column** contains information about the same columns of the aforementioned sequences, and each **row** matches one amino acid. In this manner, each **cell** points to a specific column of all observed sequences, as well as one amino acid : the value it contains is the frequency of this amino acid within these columns. PSSMs can be used to align sequences with, the result indicating whether that sequence contains a subsequence that's similar to all of the sequences represented by the PSSM.\n",
    "\n",
    "### WebLogo\n",
    "\n",
    "One way to represent a profile such as a PSSM is a table containing each value from the matrix. A *WebLogo* is a more human-readable representation, where amino acids that appear in each column are represented sequencially and with sizes depending on their frequency within that column. Here is an example of a WebLogo, gathered from [this site](http://weblogo.threeplusone.com/examples.html):\n",
    "![title](images/weblogo-example.PNG)\n",
    "\n",
    "\n",
    "### Implementation\n",
    "\n",
    "The eponym class `PSSM` contains the frequency matrix as well as metadata that will allow us to use it efficiently as a scoring system when aligning sequences with it (more on that later). Here is the class, as well as a dictionary of absolute amino acid frequencies within the UniProt database (also used to provide scores) :"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scoring\n",
    "\n",
    "Suppose we want to align a sequence to a PSSM: it will be the PSSM that determines the scores for the alignment -based on the frequencies gathered from previously aligned sequences-. Sure, PSSM has \"Scoring Matrix\" in it, but we haven't talked about scores yet (that is, unless you've read the code above). That's when you should think : \"oh, but I know what a scoring matrix is, it gives out scores for each amino acid pair, that's what this is\". Well, when we were aligning two sequences together, that was the case since we made the assumption that the positions of amino acids didn't matter. However the goal now is to align one sequence with a load of other aligned sequences -represented by a PSSM- that often don't have the same amino acid at the same position. The frequency and range of amino acids can vary for each column of the PSSM, therefore the column value is required (instead of the \"other\" amino acid) to provide a score.\n",
    "\n",
    "So how is that score calculated you ask ? Well, exactly the same way we did for the Blosum matrices: with the **log-odds ratio** $\\frac{q_{a,c}}{p_a}$, where $q_{a,c}$ is the evolutionary probability of amino acid $a$ being located at column $c$, and $p_a$ is the random probability of amino acid $a$ (which doesn't depend on the column). What differs from the previous project is the way these terms are calculated:\n",
    "* The **evolutionary probability** is computed as follows: $$q_{a,c} = \\frac{\\alpha f_{a,c} + \\beta p_{a}}{\\alpha + \\beta}$$ where $f_{a,c}$ is the frequency of $a$ in column $c$, and $\\alpha$ and $\\beta$ are constants -called *pseudocounts*- used to prevent null probabilities from happening (since $\\log{0} = - \\infty$ isn't a nice value for a score). Pseudocounts not only avoid bad values, they determine the PSSM before any aligned sequences are added. Should there be no amino acids in any column, the score would entirely depend on these pseudocounts. Here, $\\alpha$ is the total number of amino acids in the given column (without taking gaps into account) and $\\beta$ is the total number of aligned sequences in the PSSM.\n",
    "* The **random probability** is based on the frequencies of amino acids in the whole UniProt database, which can be found [here](http://web.expasy.org/docs/relnotes/relstat.html).\n",
    "\n",
    "\n",
    "### Alignment\n",
    "\n",
    "This is going to be easy: in order to align a sequence with a PSSM, we can re-use the algorithm written for the first project to compute the alignment. The code has been adapted to better manage and represent MSAs, but the alignment is basically a local+suboptimal sequence alignment, with no affine gap penalty, where scores as well are provided by the PSSM. Note that gap penalties can be made to depend on the column and therefore are provided by the PSSM class, however in this case we use a constant gap penalty throug the whole alignment.\n",
    "\n",
    "## Results\n",
    "\n",
    "### The WW domain\n",
    "\n",
    "Here is the HMM Logo (similar to a WebLogo) of the WW domain, provided by [this site](http://pfam.xfam.org/family/ww#tabview=tab4) (note that columns with no \"main\" amino acid are not represented):\n",
    "![title](images/weblogo-WW-xfam.PNG)\n",
    "\n",
    "Once aligned with the two mentioned online tools, we can represent our human proteins-only WW domains with WebLogos directly from [this site](http://weblogo.threeplusone.com/create.cgi). Here are the results of the CLUSTAL Omega alignment:\n",
    "![title](images/weblogo-WW-CLUSTALO.PNG)\n",
    "\n",
    "and the MUSCLE alignemnt:\n",
    "![title](images/weblogo-WW-MUSCLE.png)\n",
    "\n",
    "Despite the ranges being slightly different between the WW domain and our subset aligned \"by hand\", the similarities speak for themselves: the same amino acids are of great importance to the domain, only \"smaller\" amino acids differ. Conserved positions are mainly 3,6,18,19,54,57.\n",
    "\n",
    "### Sequence alignment\n",
    "\n",
    "#### First protein\n",
    "\n",
    "Protein [D6C652](http://www.uniprot.org/uniprot/D6C652) (Transcriptional coactivator YAP1-A) domains according to UniProt:\n",
    "![title](images/D6C652-Domains.PNG)\n",
    "\n",
    "This tells us that protein contains two WW domains, starting at positions 141 and 199. Let's see if these match our results for its alignment against our PSSMs for the two MSAs of the WW domain (note that by default, the function `multiAlign` only returns one result per optimal or suboptimal lookup, and only goes to suboptimal depth of 3):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " >>> Creating PSSM from file ..\\resources\\fasta\\msaresults-MUSCLE.fasta ... done\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 59\n",
      "Type       : local\n",
      "Score      : 25.92\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t28 Gaps, 31 AAs (positions 142 to 173)\n",
      "\n",
      "142\n",
      "-LPPGWEMAKT-PS-GQR-YFLN------------------------HIDQTTTWQDPR\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local-suboptimal(1)\n",
      "Score      : 23.73\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t28 Gaps, 32 AAs (positions 200 to 232)\n",
      "\n",
      "200\n",
      "-LPDGWEQALTPEGEA---YFIN------------------------HKNKSTSWLDPRL\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 5\n",
      "Type       : local-suboptimal(2)\n",
      "Score      : 11.06\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t1 Gaps, 4 AAs (positions 254 to 258)\n",
      "\n",
      "254\n",
      "-LPPP\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local-suboptimal(3)\n",
      "Score      : 10.73\n",
      "Gaps       : 41\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t41 Gaps, 19 AAs (positions 121 to 140)\n",
      "\n",
      "121\n",
      "-LA-P-------PSA--P-----------------------------HL-RQSSYEIPDD\n",
      "\n",
      "\n",
      "\n",
      " >>> Creating PSSM from file ..\\resources\\fasta\\msaresults-CLUSTALO.fasta ..."
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '..\\\\resources\\\\fasta\\\\msaresults-CLUSTALO.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-ef1f23f5bc67>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"\\n >>> Creating PSSM from file {} ...\"\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maligned\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mend\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mpssm\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mPSSM\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"WW domain\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[1;32mfor\u001b[0m \u001b[0mseq\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mloadFasta\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maligned\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m                 \u001b[0mpssm\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0mpssm\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msetGapPenalty\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\sequence.py\u001b[0m in \u001b[0;36mloadFasta\u001b[0;34m(path)\u001b[0m\n\u001b[1;32m    215\u001b[0m         \u001b[0mLoads\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mfasta\u001b[0m \u001b[0mfile\u001b[0m \u001b[0mlocated\u001b[0m \u001b[1;32min\u001b[0m \u001b[1;34m'path'\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0myields\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mSequences\u001b[0m \u001b[0mit\u001b[0m \u001b[0mcontains\u001b[0m\u001b[1;33m.\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    216\u001b[0m \t\"\"\"\n\u001b[0;32m--> 217\u001b[0;31m         \u001b[1;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mfastaFile\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    218\u001b[0m                 \u001b[0mnewSequence\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m                 \u001b[1;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mfastaFile\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '..\\\\resources\\\\fasta\\\\msaresults-CLUSTALO.fasta'"
     ]
    }
   ],
   "source": [
    "musclePath = normpath(\"../resources/fasta/msaresults-MUSCLE.fasta\")\n",
    "clustalPath = normpath(\"../resources/fasta/msaresults-CLUSTALO.fasta\")\n",
    "testPath = normpath(\"../resources/fasta/test.fasta\")\n",
    "\n",
    "for aligned in (musclePath, clustalPath):\n",
    "\tprint(\"\\n >>> Creating PSSM from file {} ...\".format(aligned), end=\"\")\n",
    "\tpssm = PSSM(\"WW domain\")\n",
    "\tfor seq in loadFasta(aligned):\n",
    "\t\tpssm.add(seq)\n",
    "\tpssm.setGapPenalty(4)\n",
    "\tprint(\" done\\n\\n\")\n",
    "\t\n",
    "\t\n",
    "\tal = Align(pssm)\n",
    "\ttoalign = [p for p in loadFasta(testPath)][0] #First protein\n",
    "\tfor aligned in al.multiAlign(toalign):\n",
    "\t\tprint(aligned)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* For the **MUSCLE** PSSM, the first two alignments match the ones from UniProt, and the last two are lesser alignments (with lower scores) that do not seem to represent a domain -mainly because of their reduced size or number of amino acids-.\n",
    "* For the **CLUSTAL** PSSM, the first two alignments match part of one domain, the last two part of the other. This may be due to different PSSM values and/or non-ideal end conditions for the align algorithm (which could be modified to prefer longer alignments). This matter could be settled by examining all outputs (not one per backtrack) and determining if and when better domain alignments are returned.\n",
    "\n",
    "#### Second protein\n",
    "\n",
    "Protein [P46935](http://www.uniprot.org/uniprot/P46935) (E3 ubiquitin-protein ligase NEDD4) domains according to UniProt:\n",
    "![title](images/P46935-Domains.PNG)\n",
    "\n",
    "Let's see how these correlate to our own results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for aligned in (musclePath, clustalPath):\n",
    "\tprint(\"\\n >>> Creating PSSM from file {} ...\".format(aligned), end=\"\")\n",
    "\tpssm = PSSM(\"WW domain\")\n",
    "\tfor seq in loadFasta(aligned):\n",
    "\t\tpssm.add(seq)\n",
    "\tpssm.setGapPenalty(4)\n",
    "\tprint(\" done\\n\\n\")\n",
    "\t\n",
    "\t\n",
    "\tal = Align(pssm)\n",
    "\ttoalign = [p for p in loadFasta(testPath)][1] #Second protein\n",
    "\tfor aligned in al.multiAlign(toalign):\n",
    "\t\tprint(aligned)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* For the **MUSCLE** PSSM, the three instances of the WW domain are found within the four results. One alignment is found to have a better score than the last WW domain, yet it is included within the latter. This may be due to the align algorithm once more, which could be tweaked to prefer longer alignments even when this means not starting at the maximum score.\n",
    "* For the **CLUSTAL** PSSM, no instances of the WW domain are found completely. Yet all of the results are subsets of the domains, which leads to the same conclusions as before.\n",
    "\n",
    "## Conclusion\n",
    "\n",
    "The PSSM seems to work exactly like it should, providing the required information to find local alignments within other sequences. By looking at the WebLogos, we can see that both online tools have aligned our sequences from the WW domain in a very similar fashion and resemble the HMM Logo of the WW domain, yet the MUSCLE alignment provides better results than the OMEGA alignemnt. \n",
    "\n",
    "All of the WW domains from both aligned sequences have been found, however the algorithm can probably be tweaked to have better start and end conditions for the alignment, since it sometimes returns only part of the domain. We can also notice that, when a complete domain is found, its start and end index within the sequence differ very slighlty (one or two amino acids) with the UniProt alignment; this could be due to tiny differences in the PSSM values, to the arbitrary selection of results from the backtracking (only the first is displayed for each lookup), or even to the constant and arbitrary gap penalty (different gap penalties per column have been implemented but not used).\n",
    "\n",
    "Considering that the two tested sequences come from different organisms than the ones used for creating the PSSM, the results are positive overall."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
