{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from sys import path\n",
    "from os.path import join as joinpath\n",
    "from os.path import normpath\n",
    "path.append(normpath(\"../src\"))\n",
    "\n",
    "from aminoacid import AminoAcid\n",
    "from sequence import Sequence, loadFasta\n",
    "from score import ScoreMatrix\n",
    "from align import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence alignment\n",
    "\n",
    "Genetic material such as DNA and proteins comes in sequences made up of different blocks. The study of these sequences and their correlation is invaluable to the field of biology. Since genetic material is likely to mutate over time, it is not easy to be certain of these correlations in most cases. In order to know if two or more sequences have a common origin, we must examine their similarity in a formal way, and be able to quantify it. Since the modifications of the sequences can introduce new blocks as well as delete some, the first step to comparing sequences is aligning them by making the similar regions match. Even then, high levels of similarity do not necessarily imply a common ancestor exists (homology). Our goal in this section is to create a tool that allows us to perform such alignments and give a numeric value of the chance of homology.\n",
    "\n",
    "\n",
    "## ADTs\n",
    "\n",
    "We will first look at an implementation of Abstract Data Types (ADTs) that will allow us to represent Sequences. \n",
    "\n",
    "### AminoAcid\n",
    "\n",
    "The choice of genetic material is proteins. In the case of proteins, the blocks that make up the sequences are called amino acids. Therefore, they are the first thing we must be able to represent. The following tuple lists all amino acids and miscellaneous options.\n",
    "\n",
    "The following class allows us to create AminoAcid objects that represent any of the amino acids from the list. We can then compare, print, hash, test them with any of the defined methods. Do note that we can create them by copying other AminoAcids or by providing their name in any form (long for complete, medium for 3 chars or short for 1 char).\n",
    "\n",
    "Let's test it a little bit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "True\n",
      "lysine\n"
     ]
    }
   ],
   "source": [
    "a1 = AminoAcid(\"A\")\n",
    "print(a1)\n",
    "a2 = AminoAcid(a1)\n",
    "print(a1 == a2)\n",
    "a3 = AminoAcid(\"K\")\n",
    "print(a3.getName(\"long\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence\n",
    "\n",
    "That's good and all, but we don't want to align single amino acids do we ? What we want to align is sequences. We can view a Sequence as an ordered list of AminoAcid objects. It should be iterable : we'll want to go through, access, and count its items. We'll also want to change them, insert new ones, delete some, and do all that transparently, as we would with a list. Finally it would be useful to check whether a sequence contains some other sub-sequence or single amino acid. The following is a class that does just that.\n",
    "\n",
    "Since we may not want to type every sequence we use by hand, we can also read them from files. The format chosen here is .fasta, but this can be adapted for any other format.\n",
    "\n",
    "Let's test a few of the capabilities of this class :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABX\n",
      "ABXC\n",
      "True\n",
      "lysine-cysteine, lysine-cysteine, ABXC\n",
      "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
      "26\n"
     ]
    }
   ],
   "source": [
    "s = Sequence(\"ABX\")\n",
    "print(s)\n",
    "s.extend(\"cysteine\")\n",
    "print(s)\n",
    "print(AminoAcid(\"alanine\") in s)\n",
    "sCopy = Sequence(s) #This is a deep copy\n",
    "sAlias = s #This is the same object\n",
    "del s[1:3]\n",
    "sAlias[0] = \"K\"\n",
    "s.setSeparator(\"-\")\n",
    "s.setNameMode(\"long\")\n",
    "print(s, sAlias, sCopy, sep=\", \")\n",
    "\n",
    "sequences = [seq for seq in loadFasta(normpath(\"../resources/fasta/SH3-sequence.fasta\"))]\n",
    "print(sequences[0])\n",
    "print(sequences[0].hasSubSequence(Sequence(\"RLQIVNNTEGDWWLAHSLS\"))) #Returns the index of the sub-sequence, if found"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Score\n",
    "\n",
    "Now that the easy part is done, we need to remember why we're here (it's sequence alignment, just in case). We can align two sequences any number of ways, but we're only interested in alignments that could represent two descendants of the same original sequence. Therefore, we need to assign a **score** value to each alignment, that will help us see how significative it is.\n",
    "\n",
    "Because of the way mutations happen, all modifications should not be treated equally. Some amino acids are more related between them than others, therefore making them more likely to mutate into each other. Some changes are more or less likely to happen because of the very structure of the protein. Also, when an amino acid is inserted into or deleted from a sequence (thus creating a **gap** in the alignment), it usually is not the only one. This tells us that, not only should the score of the alignment depend on each pair that's aligned (meaning, on the same location), but it should also depend on how many gaps there are and how big they are.\n",
    "\n",
    "The `Score` class allows us to give a numerical value to each possible pair of amino acids. These values can be set manually one by one, just as they can be loaded from files. The chosen format here was `.iij`.\n",
    "\n",
    "Here's an example on how to load a `Score` object from a `.iij` file, and access the scores for certains pairs :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- BLOSUM 62 ----------\n",
      "A   4   \n",
      "R   -1  5   \n",
      "N   -2  0   6   \n",
      "D   -2  -2  1   6   \n",
      "C   0   -3  -3  -3  9   \n",
      "Q   -1  1   0   0   -3  5   \n",
      "E   -1  0   0   2   -4  2   5   \n",
      "G   0   -2  0   -1  -3  -2  -2  6   \n",
      "H   -2  0   1   -1  -3  0   0   -2  8   \n",
      "I   -1  -3  -3  -3  -1  -3  -3  -4  -3  4   \n",
      "L   -1  -2  -3  -4  -1  -2  -3  -4  -3  2   4   \n",
      "K   -1  2   0   -1  -3  1   1   -2  -1  -3  -2  5   \n",
      "M   -1  -1  -2  -3  -1  0   -2  -3  -2  1   2   -1  5   \n",
      "F   -2  -3  -3  -3  -2  -3  -3  -3  -1  0   0   -3  0   6   \n",
      "P   -1  -2  -2  -1  -3  -1  -1  -2  -2  -3  -3  -1  -2  -4  7   \n",
      "S   1   -1  1   0   -1  0   0   0   -1  -2  -2  0   -1  -2  -1  4   \n",
      "T   0   -1  0   -1  -1  -1  -1  -2  -2  -1  -1  -1  -1  -2  -1  1   5   \n",
      "W   -3  -3  -4  -4  -2  -2  -3  -2  -2  -3  -2  -3  -1  1   -4  -3  -2  11  \n",
      "Y   -2  -2  -2  -3  -2  -1  -2  -3  2   -1  -1  -2  -1  3   -3  -2  -2  2   7   \n",
      "V   0   -3  -3  -3  -1  -2  -2  -3  -3  3   1   -2  1   -1  -2  -2  0   -3  -1  4   \n",
      "B   -2  -1  3   4   -3  0   1   -1  0   -3  -4  0   -3  -3  -2  0   -1  -4  -3  -3  4   \n",
      "Z   -1  0   0   1   -3  3   4   -2  0   -3  -3  1   -1  -3  -1  0   -1  -3  -2  -2  1   4   \n",
      "X   0   -1  -1  -1  -2  -1  -1  -1  -1  -1  -1  -1  -1  -1  -2  0   0   -2  -1  -1  -1  -1  -1  \n",
      "\n",
      "    A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V   B   Z   X   \n",
      "H N  :  1\n"
     ]
    }
   ],
   "source": [
    "scoring = ScoreMatrix(normpath(\"../resources/blosum/blosum62.iij\"), \"BLOSUM 62\")\n",
    "print(scoring)\n",
    "\n",
    "a1, a2 = Sequence(\"HN\") #That's call unpacking, pretty neat huh ?\n",
    "print(a1, a2, \" : \", scoring.getScore(a1, a2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Needleman-Wunsch Alignment\n",
    "\n",
    "I know, I know : all these lines and still no alignment in sight. What does that title even mean ? Well, in 1970 Saul B. Needleman and Christian D. Wunsch came up with an effective algorithm for aligning sequences. It provides us with the alignments that get the best score, given certain conditions. It uses a scoring system like the one we've just covered, with the addition of gap penalties : negative values added to the score when the alignment creates (initial penalty) and extends (extended penalty) a gap. Here's an overview of its steps :\n",
    " * Create a matrix with enough rows to fit one sequence ($A$) and enough columns to fit the other ($B$) : each cell $(i, j)$ represents a possible alignment between two amino acids $A_i$ and $B_j$.\n",
    " * Add an initial row and column to the matrix, with values (scores) determined a certain number of ways. Keep in mind that these cells represent the beginning of an alignment where one sequence only has gaps, same with the last row and column for the end of the alignment.\n",
    " * Go through every cell in the matrix and calculate its score based on the previous (left, top, left and top) cells, using the following formula where $(V,W,S)_{i,j}$ are 3 values contained in cell $(i,j)$ of the matrix, $Score(A_i, B_j)$ is the score between amino acids $A_i$ and $B_j$, $I$ is the initial gap penalty and $E$ the extended gap penalty.\n",
    "$$\n",
    "V(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i-1, j) + I\\\\\n",
    "      V(i-1, j) - E\n",
    "    \\end{array}\n",
    "\\right.\n",
    "\\quad\n",
    "W(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i, j-1) + I\\\\\n",
    "      V(i, j-1) - E\n",
    "    \\end{array}\n",
    "\\right.\n",
    "\\quad\n",
    "S(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i-1, j-1) + Score(A_i, B_j)\\\\\n",
    "      V(i, j)\\\\\n",
    "      W(i, j)\n",
    "    \\end{array}\n",
    "\\right.\n",
    "$$\n",
    " * Backtrack from some point of the matrix (the end of the alignment) to some other (the beginning), only passing by permitted cells. The cells allowed after cell $(i,j)$ are the ones where the value $S(i,j)$ comes from : \n",
    "     * left if $S(i,j)=W(i,j)$ : sequence $A$ has a gap\n",
    "     * top if $S(i,j)=V(i,j)$ : sequence $B$ has a gap\n",
    "     * diagonal if $S(i,j)=S(i-1,j-1) + Score(A_i, B_j)$ : sequences $A$ and $B$ are aligned\n",
    "     \n",
    "Different types of alignments can be obtained by tweaking this algorithm.\n",
    " * **Global alignments** aim to align both sequences completely. In order to do that, we initialize the first row and sequences with multiples of $I$ and $E$, thus giving us negative values matching the gap required to get there. Backtracking starts at the end of the matrix and ends at the beginning.\n",
    " * **Local alignments** aim to produce the alignment with the best score, whithout regard for their lenght. We do not initialize the first row and column : completing an alignment with only gaps has no interest score-wise. Backtracking starts at the highest value(s) in the matrix and ends as soon as we reach a value of $0$. Local **suboptimal alignments** can be found by clearing the values of the local optimal alignment in the matrix, reevaluating scores for further rows and columns, and backtracking again.\n",
    " * **Semiglobal alignments** are intended for a global-like alignment between sequences that only overlap partially, or that have great difference in size (one is included in the other). The first row and column are not initialized for the same reason as with local alignments. Backtracking starts and tha highest value(s) but ends when we reach either the first line or first column (therefore finishing one sequence).\n",
    "\n",
    "\n",
    "The following is a class that allows us to represent two aligned sequences, along with information about the way they were aligned and the result. **Identity** is the number of equal amino acids that are aligned, **similarity** is the number of similar (meaning equal or with a non negative score) amino acids that are aligned.\n",
    "\n",
    "In order to use this class, we must (finally) implement the actual alignment algorithm. This is done by the following class.\n",
    "\n",
    "Here is the result of a global alignment between the first two sequences from \"SH3-sequences.fasta\", calculated by LALIGN with the BLOSUM62 scoring matrix, initial and extended gap penalties of $-12$ and $-2$ :\n",
    "```\n",
    "n-w opt:  69  Z-score: 152.4  bits: 31.6 E(1): 6.7e-25\n",
    "global/global (N-W) score: 69; 29.0% identity (62.9% similar) in 62 aa overlap (1-62:1-58)\n",
    "\n",
    "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
    "      .: ::... .. .::::.:. :...:.   . :   .:. :. :.::.::.  .\n",
    "---MEAIAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELN-GKDGFIPKNYIEMKPH\n",
    "```\n",
    "\n",
    "Next is the result from the AlignMatrix class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Size       : 62\n",
      "Type       : global\n",
      "Score      : 73\n",
      "Identity   : 18 (29.03%)\n",
      "Similarity : 21 (33.87%)\n",
      "Gaps       : 4\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 62 AAs (positions 0 to 62)\n",
      "Lower seq. : sp|P62993|1-58\n",
      "\t4 Gaps, 58 AAs (positions 0 to 58)\n",
      "\n",
      "0\n",
      "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
      "      .: ::... .. .::::.:. :...:.   . :   .:. :. :.::.::.  .  \n",
      "0\n",
      "---MEAIAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELN-GKDGFIPKNYIEMKPH\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = Align(scoring)\n",
    "\n",
    "for align in a.globalAlign(sequences[0], sequences[1], -12, -2, False):\n",
    "\tprint(align)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Besides the second one being much more readable, their output is pretty similar. There may be slight differences in the BLOSUM matrix used, responsible for the discrepancy between the scores.\n",
    "\n",
    "Here is the result of a local alignment between the first two sequences from \"maguk-sequences.fasta\", calculated by LALIGN with the BLOSUM62 scoring matrix, initial and extended gap penalties of $-12$ and $-2$ :\n",
    "```\n",
    "Waterman-Eggert score: 2677;  1048.2 bits; E(1) <  0\n",
    "69.5% identity (86.3% similar) in 767 aa overlap (153-904:61-817)\n",
    "\n",
    "SHSHISPIKPTEA-VLPSPPTVPVIPVLPVPAENTVILP-TIPQANPPPVLVNTDSLETP\n",
    ":..  .:   :.: ..:.   :  .:  :::...:  :  . :.  :  . .: :  :  \n",
    "SQAGATPTPRTKAKLIPTGRDVGPVPPKPVPGKSTPKLNGSGPSWWPECTCTNRDWYEQ-\n",
    "\n",
    "TYVNGTDADYEYEEITLERGNSGLGFSIAGGTDNPHIGDDSSIFITKIITGGAAAQDGRL\n",
    "  :::.:. ..::::.::::::::::::::: ::::. :: .::::::: :::::.::::\n",
    "--VNGSDGMFKYEEIVLERGNSGLGFSIAGGIDNPHVPDDPGIFITKIIPGGAAAMDGRL\n",
    "\n",
    "RVNDCILRVNEVDVRDVTHSKAVEALKEAGSIVRLYVKRRKPVSEKIMEIKLIKGPKGLG\n",
    " ::::.:::::::: .:.::.::::::::: .::: :.::.:  : :::..:.:::::::\n",
    "GVNDCVLRVNEVDVSEVVHSRAVEALKEAGPVVRLVVRRRQPPPETIMEVNLLKGPKGLG\n",
    "\n",
    "FSIAGGVGNQHIPGDNSIYVTKIIEGGAAHKDGKLQIGDKLLAVNNVCLEEVTHEEAVTA\n",
    "::::::.::::::::::::.:::::::::.:::.:::::.::::::. :..: :::::..\n",
    "FSIAGGIGNQHIPGDNSIYITKIIEGGAAQKDGRLQIGDRLLAVNNTNLQDVRHEEAVAS\n",
    "      \n",
    "LKNTSDFVYLKVAKPTSMYMNDGYAPPDITNSSSQPVDNHVSPSSFLGQTPA--------\n",
    "::::::.:::::::: :...:: ::::: ... .  .:::.: .: ::   :        \n",
    "LKNTSDMVYLKVAKPGSLHLNDMYAPPDYASTFTALADNHISHNSSLGYLGAVESKVSYP\n",
    "     \n",
    "-----SPARYSPVSKAVLGDDEITREPRKVVLHRGSTGLGFNIVGGEDGEGIFISFILAG\n",
    "      :.::::. . .:.....::::::..::.:::::::::::::::::::.::::::\n",
    "APPQVPPTRYSPIPRHMLAEEDFTREPRKIILHKGSTGLGFNIVGGEDGEGIFVSFILAG\n",
    "  \n",
    "GPADLSGELRKGDRIISVNSVDLRAASHEQAAAALKNAGQAVTIVAQYRPEEYSRFEAKI\n",
    "::::::::::.::::.:::.:.:: :.:::::::::.:::.::::::::::::::::.::\n",
    "GPADLSGELRRGDRILSVNGVNLRNATHEQAAAALKRAGQSVTIVAQYRPEEYSRFESKI\n",
    "    \n",
    "HDLREQMMNSSISSGSGSLRTSQKRSLYVRALFDYDKTKDSGLPSQGLNFKFGDILHVIN\n",
    ":::::::::::.::::::::::.:::::::::::::.:.:: ::::::.:..::::::::\n",
    "HDLREQMMNSSMSSGSGSLRTSEKRSLYVRALFDYDRTRDSCLPSQGLSFSYGDILHVIN\n",
    "   \n",
    "ASDDEWWQARQVTPDGESDEVGVIPSKRRVEKKERARLKTVKFNSKTRDKGEIPDDMGSK\n",
    ":::::::::: ::: :::...::::::.:::::::::::::::...:   : : ..    \n",
    "ASDDEWWQARLVTPHGESEQIGVIPSKKRVEKKERARLKTVKFHART---GMIESNRDFP\n",
    "   \n",
    "GLKHVTSNASDSESSYRGQEEYVLSYEPVNQQEVNYTRPVIILGPMKDRINDDLISEFPD\n",
    "::    :.   . .. .:::. .::::::..::..:.::::::::::::.::::::::: \n",
    "GL----SDDYYGAKNLKGQEDAILSYEPVTRQEIHYARPVIILGPMKDRVNDDLISEFPH\n",
    "  \n",
    "KFGSCVPHTTRPKRDYEVDGRDYHFVTSREQMEKDIQEHKFIEAGQYNNHLYGTSVQSVR\n",
    "::::::::::::.:: ::::.:::::.::::::::::..:::::::.:..:::::.::::\n",
    "KFGSCVPHTTRPRRDNEVDGQDYHFVVSREQMEKDIQDNKFIEAGQFNDNLYGTSIQSVR\n",
    "    \n",
    "EVAEKGKHCILDVSGNAIKRLQIAQLYPISIFIKPKSMENIMEMNKRLTEEQARKTFERA\n",
    " :::.::::::::::::::::: ::::::.:::::::.: .::::.: : :::.: ...:\n",
    "AVAERGKHCILDVSGNAIKRLQQAQLYPIAIFIKPKSIEALMEMNRRQTYEQANKIYDKA\n",
    " \n",
    "MKLEQEFTEHFTAIVQGDTLEDIYNQVKQIIEEQSGSYIWVPAKEKL\n",
    "::::::: :.::::::::.::.:::..:::::.::: :::::. :::\n",
    "MKLEQEFGEYFTAIVQGDSLEEIYNKIKQIIEDQSGHYIWVPSPEKL  \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Size       : 51\n",
      "Type       : local\n",
      "Score      : 98\n",
      "Identity   : 18 (35.29%)\n",
      "Similarity : 20 (39.22%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 51 AAs (positions 6 to 57)\n",
      "Lower seq. : sp|P62993|1-58\n",
      "\t1 Gaps, 50 AAs (positions 3 to 53)\n",
      "\n",
      "6\n",
      "VALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYV\n",
      ".: ::... .. .::::.:. :...:.   . :   .:. :. :.::.::.\n",
      "3\n",
      "IAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELN-GKDGFIPKNYI\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for aligned in a.localAlign(sequences[0], sequences[1], -12, -2):\n",
    "\taligned.chunkSize = 60\n",
    "\tprint(aligned)\n",
    "\tbreak"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once more, alignments are quite similar in terms of gap locations, scores and identity/similarity. The results suggest that these sequences might be related, given their high identity and similarity.\n",
    "\n",
    "For each other pair of sequences (condensed results), we don't get such high changes of homology, as suggested by these condensed results :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Size       : 51\n",
      "Type       : local\n",
      "Score      : 98\n",
      "Identity   : 18 (35.29%)\n",
      "Similarity : 20 (39.22%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 51 AAs (positions 6 to 57)\n",
      "Lower seq. : sp|P62993|1-58\n",
      "\t1 Gaps, 50 AAs (positions 3 to 53)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 54\n",
      "Type       : local\n",
      "Score      : 93\n",
      "Identity   : 21 (38.89%)\n",
      "Similarity : 13 (24.07%)\n",
      "Gaps       : None\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 54 AAs (positions 3 to 57)\n",
      "Lower seq. : sp|P41240|9-70\n",
      "\t0 Gaps, 54 AAs (positions 3 to 57)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 51\n",
      "Type       : local\n",
      "Score      : 81\n",
      "Identity   : 17 (33.33%)\n",
      "Similarity : 18 (35.29%)\n",
      "Gaps       : 2\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 51 AAs (positions 6 to 57)\n",
      "Lower seq. : sp|Q8IZP0|446-505\n",
      "\t2 Gaps, 49 AAs (positions 6 to 55)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 55\n",
      "Type       : local\n",
      "Score      : 118\n",
      "Identity   : 24 (43.64%)\n",
      "Similarity : 19 (34.55%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P12931|84-145\n",
      "\t0 Gaps, 55 AAs (positions 6 to 61)\n",
      "Lower seq. : sp|Q06187|214-274\n",
      "\t1 Gaps, 54 AAs (positions 6 to 60)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 56\n",
      "Type       : local\n",
      "Score      : 135\n",
      "Identity   : 25 (44.64%)\n",
      "Similarity : 18 (32.14%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P62993|1-58\n",
      "\t1 Gaps, 55 AAs (positions 1 to 56)\n",
      "Lower seq. : sp|P41240|9-70\n",
      "\t0 Gaps, 56 AAs (positions 4 to 60)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 57\n",
      "Type       : local\n",
      "Score      : 107\n",
      "Identity   : 22 (38.60%)\n",
      "Similarity : 17 (29.82%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P62993|1-58\n",
      "\t0 Gaps, 57 AAs (positions 1 to 58)\n",
      "Lower seq. : sp|Q8IZP0|446-505\n",
      "\t1 Gaps, 56 AAs (positions 4 to 60)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 53\n",
      "Type       : local\n",
      "Score      : 90\n",
      "Identity   : 18 (33.96%)\n",
      "Similarity : 20 (37.74%)\n",
      "Gaps       : 2\n",
      "\n",
      "Upper seq. : sp|P62993|1-58\n",
      "\t1 Gaps, 52 AAs (positions 1 to 53)\n",
      "Lower seq. : sp|Q06187|214-274\n",
      "\t1 Gaps, 52 AAs (positions 4 to 56)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 52\n",
      "Type       : local\n",
      "Score      : 62\n",
      "Identity   : 13 (25.00%)\n",
      "Similarity : 20 (38.46%)\n",
      "Gaps       : 2\n",
      "\n",
      "Upper seq. : sp|P41240|9-70\n",
      "\t0 Gaps, 52 AAs (positions 6 to 58)\n",
      "Lower seq. : sp|Q8IZP0|446-505\n",
      "\t2 Gaps, 50 AAs (positions 6 to 56)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 55\n",
      "Type       : local\n",
      "Score      : 94\n",
      "Identity   : 21 (38.18%)\n",
      "Similarity : 15 (27.27%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|P41240|9-70\n",
      "\t0 Gaps, 55 AAs (positions 6 to 61)\n",
      "Lower seq. : sp|Q06187|214-274\n",
      "\t1 Gaps, 54 AAs (positions 6 to 60)\n",
      "\n",
      "---------- Alignment ----------\n",
      "Size       : 54\n",
      "Type       : local\n",
      "Score      : 73\n",
      "Identity   : 16 (29.63%)\n",
      "Similarity : 21 (38.89%)\n",
      "Gaps       : 1\n",
      "\n",
      "Upper seq. : sp|Q8IZP0|446-505\n",
      "\t1 Gaps, 53 AAs (positions 2 to 55)\n",
      "Lower seq. : sp|Q06187|214-274\n",
      "\t0 Gaps, 54 AAs (positions 2 to 56)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(sequences)-1):\n",
    "\tfor j in range(i+1, len(sequences)):\n",
    "\t\tfor align in a.localAlign(sequences[i], sequences[j], -12, -2):\n",
    "\t\t\talign.condensed = True\n",
    "\t\t\tprint(align)\n",
    "\t\t\tbreak"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
